pkg/lightning/backend/external/engine.go

```diff
@@ -37,6 +37,38 @@ import (
 	"go.uber.org/zap"
 )
 
+// writeStepMemShareCount defines the number of shares of memory per job worker.
+// For each job worker, the memory it can use is determined by cpu:mem ratio, say
+// a 16c32G machine, each worker can use 2G memory.
+// And for the memory corresponding to each job worker, we divide into below and
+// total 6.5 shares:
+//   - one share used by HTTP and GRPC buf, such as loadBatchRegionData, write TiKV
+//   - one share used by loadBatchRegionData to store loaded data batch A
+//   - one share used by generateAndSendJob for handle loaded data batch B
+//   - one share used by the active job on job worker
+//   - 2.5 share for others, and burst allocation to avoid OOM
+//
+// the share size 'SS' determines the max range data size 'RangeS' for a range job.
+// Each range job corresponding to one ingested SST on TiKV, we want to load as
+// many data as possible, while minimizing the number of SSTs (too many SST file,
+// say 500K, will cause TiKV slow down when ingest), and make the SST be more even,
+// so we calculate RangeS by:
+//   - let TempRangeS = SS
+//   - if TempRangeS < RS, RangeS = RS / ceil(RS / TempRangeS) + 1,
```

Comment:
RS is region size?

Reply:
yes, i will add it later

Reply:
https://github.com/pingcap/tidb/pull/59870/commits/c5a2f78322027bf09a66e62e412552b3178a98a5